crates/cairo-serde-derive/src/derive_enum.rs

```diff
@@ -0,0 +1,198 @@
+use proc_macro2::{Span, TokenStream};
+use quote::quote;
+use syn::{DataEnum, Ident, Type, Variant};
+use unzip_n::unzip_n;
+
+pub fn derive_enum(ident: Ident, data: DataEnum) -> TokenStream {
+    let matches = &data
+        .variants
+        .iter()
+        .map(|v| derive_enum_matches(&ident, v))
+        .collect::<Vec<_>>();
+
+    unzip_n!(3);
+    let (serialized_size, serialize, deserialize) = data
+        .variants
+        .iter()
+        .enumerate()
+        .map(|(i, v)| derive_enum_variant(&ident, i, v))
+        .collect::<Vec<_>>()
+        .into_iter()
+        .unzip_n_vec();
+
+    let cairo_serialized_size = quote! {
+        fn cairo_serialized_size(rust: &Self::RustType) -> usize {
+            match rust {
+                #(
+                    #matches => #serialized_size,
+                )*
+            }
+        }
+    };
+
+    let cairo_serialize = quote! {
+        fn cairo_serialize(rust: &Self::RustType) -> Vec<::starknet::core::types::Felt> {
+            match rust {
+                #(
+                    #matches => #serialize,
+                )*
+            }
+        }
+    };
+
+    let deserialize_matches = data
+        .variants
+        .iter()
+        .enumerate()
+        .map(|(i, _)| syn::LitInt::new(&i.to_string(), Span::call_site()))
+        .collect::<Vec<_>>();
+    let cairo_deserialize = quote! {
+        fn cairo_deserialize(felt: &[Felt], offset: usize) -> Result<Self::RustType, ::cainome_cairo_serde::Error> {
+            let offset = offset + 1;
+            #(
+                if felt[offset - 1] == ::starknet::core::types::Felt::from(#deserialize_matches) {
+                    return Ok(#deserialize);
+                }
+            )*
+            Err(::cainome_cairo_serde::Error::Deserialize("Invalid variant Id".to_string()))
```

Comment:
```suggestion
        Err(::cainome_cairo_serde::Error::Deserialize(format!(
            "Invalid variant Id `{}` deserializing {} at offset `{}`",
            felt[offset - 1],
            #ident,
            offset - 1
        )))
```
As we have the info, we should try to give more context in case this unlikely situation occurs. 👍